core/src/main/scala/akka/kafka/internal/DefaultProducerStage.scala

```diff
@@ -54,11 +55,40 @@ private class DefaultProducerStageLogic[K, V, P, IN <: Envelope[K, V, P], OUT <:
   private lazy val decider: Decider =
     inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
   protected val awaitingConfirmation = new AtomicInteger(0)
+  protected var producer: Producer[K, V] = _
   private var inIsClosed = false
   private var completionState: Option[Try[Done]] = None
 
   override protected def logSource: Class[_] = classOf[DefaultProducerStage[_, _, _, _, _]]
 
+  private def producerArrived: AsyncCallback[Producer[K, V]] = getAsyncCallback(assignProducer)
+
+  protected def assignProducer(p: Producer[K, V]): Unit = {
+    producer = p
+    resumeDemand()
+  }
+
+  override def preStart(): Unit = {
+    super.preStart()
+    val producerFuture = producerFactory(materializer.executionContext)
+    if (producerFuture.isCompleted) {
```

Comment:
reove the check and put the else in the None branch of the pattern match?

Reply:
Did see that, much better.